hdr_html_canvas_element.md

```diff
@@ -283,6 +280,57 @@ the Content Light Level Information chunk can provide more accurate information
 than the maximum luminance parameter of the Mastering Display Color Volume
 chunk.
 
+As specified below, the platform does not generally apply color volume mapping
+if the color volume of the image is smaller than that of the display.
```

Comment:
On one hand this makes sense, you want to present the content as-is when possible.

OTOH, a faithful absolute colorimetric presentation does not produce then intended appearance if the viewing environment differs.

What are "platform" and "display" in this context, do they include the OS window system?

I can think of three different viewing environments in the pipeline:
- the content colorimetry has been authored for a mastering viewing environment A
- a monitor expects a video signal to be tuned for a reference viewing environment B
- the monitor adjusts the image, based on e.g. the end user brightness and contrast knobs, to look as intended in the actual viewing environment C

BT.2100 seems to have a common reference viewing environment for both PQ and HLG, so if both content and video signal follow either of them, a simple format conversion suffices. Otherwise, something more should probably happen between A and B. That you have covered by implying the viewing environment via `PredefinedColorSpace`, I guess.

From B to C may still be a mapping.

It's also possible that a monitor would not do image adjustment, which means the video signal would need to be tuned for viewing environment C directly. I'm not quire sure how PQ mode monitors/TVs behave. Maybe that depends on the vendor and model even.

All in all, I would be a bit wary of promising "generally no mapping".

Reply:
I see multiple use cases:
1. the application provides an HLG or PQ image as-is and the platform performs additional adjustments if needed (the platform might actually perform no adjustments if, for example, the image is full screen and the display supports HLG or PQ natively)
2. (a) the platform provides to the application the nominal min/max luminance of the display, i.e. headroom, (b) the application performs tone-mapping, and (c) the platform and the display may or may not perform additional adjustments (ambient light color, user preference, etc.)
3. (a) the platform provides the min/max luminance of the display, ambient light level and color, etc., (b) the application performs all adjustments (c) the platform performs no adjustments (the display might still perform adjustments unbeknownst to the platform)

I would think that (1) should be supported no matter what.

Reply:
I don't think cases 1, 2 and 3 are really separate that much. The common thing is that the image source (application) describes the conditions of the image it has, and then the operating and window system do the best they can with it. If the application got some information in order to do a better image mapping, then the used information is part of the described image conditions. Of course, I can only speak for Wayland.

We're not really disagreeing here. All I'm saying is that "the platform does not generally apply color volume mapping
if the color volume of the image is smaller than that of the display" is a bit much to promise, and also a bit vague. If people start depending on it, you have to strike the word "generally". Otherwise the promise is moot, because it might not always hold.

We won't have that promise on Wayland. The only thing we would promise is that if you tag your image with the output's image description (color space, etc.) exactly, then  the image would be sent out in the video signal without adjustments, provided it is not blended with any other window and not a subject of compositor effects. (One reason for this is to allow applications do their own complete color management, another use case is display profiling.)

Reply:
> All I'm saying is that "the platform does not generally apply color volume mapping
if the color volume of the image is smaller than that of the display" is a bit much to promise, and also a bit vague. If people start depending on it, you have to strike the word "generally". Otherwise the promise is moot, because it might not always hold.

I agree that the wording is vague.

AFAIK the idea is that, to the extent that the platform provide information on display capabilities, the platform should be encouraged to not alter colors that fit within the display's capabilities -- unless instructed otherwise by the user, etc.

Is your point is that this is not testable and that there are too many scenarios?

> The only thing we would promise is that if you tag your image with the output's image description (color space, etc.) exactly, then the image would be sent out in the video signal without adjustments, provided it is not blended with any other window and not a subject of compositor effects. 

Could this requirement (i.e. commit to a specific output color system and not modify images that match this color system) be reasonably imposed on the Web Platform in narrow scenarios, e.g., in full screen mode?

Maybe this is too narrow a use case to care about?

Reply:
> Is your point is that this is not testable and that there are too many scenarios?

I guess that depends on how strict you want to be. If you want to be lax, then maybe explain what the goal or intention is rather than trying to set rules. It is difficult to infer intention from rules, but a well communicated intention should get everyone on the same page.

Another thing is that color volume includes luminance. I'm not sure if this means absolute or relative luminance. I'm worried that this might be recommending presentation that is not adapted to the viewing conditions.

> Could this requirement (i.e. commit to a specific output color system and not modify images that match this color system) be reasonably imposed on the Web Platform in narrow scenarios, e.g., in full screen mode?

I'm not sure I understood the question, so pardon the wall of text.

Full screen mode is necessary, if we want to relay video HDR metadata straight to the display, because the metadata affects all of the screen. For Wayland desktop compositors that is annoying to implement, because changes in display driving mode or even HDR metadata (EOTF particularly) may take a while to sink in (ha!) or even blank the display for a moment.

Hence it is much more likely that a Wayland compositor chooses one mode (and static metadata set) for driving a display, and converts everything from applications to that. A Wayland compositor may be able to use the display controller hardware to do some or all of the conversion, so it may be possible to avoid GPU work in video playback.

This also means that full screen becomes irrelevant. A compositor converts all contents from all applications to the same, um, compositing color space that is tuned for each display. Even windowed video etc. should look fine and consistent.

If an application like a browser tags its window (or a video plane) with the output's image description, it only means that the Wayland compositor can skip most of the conversions and all automatic perceptual adjustments. Composition still happens in the compositing color space, but round-tripping to that is a no-op, and may even be skipped completely. But now the burden is on the application to follow the compositor's output image description, and there are only two good reasons to do that:
- the application is generating content (e.g. a game) and can choose an arbitrary image description to target for no or little extra cost, or
- the application wants to avoid the automatic conversions of a compositor for whatever reason.

For video playback performance using the output image description for video player's window content is detrimental, because it likely forces the video player to use a GPU rendering pass for color conversion, which means there cannot be any opportunity to off-load that conversion to the display controller hardware. Hence, one should pass video frames as-is from a hardware video decoder all the way to the window system, tagging them with the video's image description.

Display controller hardware color conversions are fixed-function and highly optimized for performance and low power consumption, which makes them preferable to the power hungry GPU when at all possible.

----

Referring to the diagram in https://gitlab.freedesktop.org/pq/color-and-hdr/-/merge_requests/35 , we often cannot and might not even care to control the emitted light directly, exactly because then we also need to take viewing conditions into account ourselves. As much as it pains me to leave the actual light undefined, it makes things much simpler if we can trust that whoever receives our standardized signal also takes care of adapting it appropriately.

This dilemma is  the core when trying to support both entertainment and ICC workflows in the same framework. Display-referred "do exactly as I say" vs. scene-referred and display-referred automatically adapted to the display and viewing conditions at hand. In Wayland it seems we have actually let go of even trying to guarantee the actual light at protocol level. If an end user wants that level of control for their applications, they are required to profile (measure) their monitors and use a particular workflow.

Uhh... am I even on topic here?